### PR TITLE
Swap thresholds

### PR DESCRIPTION
**Pull request by mortelli**

### CODE REVIEW

File: swap/types.go [hunk 1]
```diff
@@ -41,9 +40,7 @@ type Cheque struct {
 
 // ChequeRequestMsg is sent from a creditor to the debitor to solicit a cheque
 type ChequeRequestMsg struct {
-	Peer       enode.ID // TODO: Why is it here right now? Potentially not needed as everything goes through peer
-	PubKey     []byte   // TODO: Also probably not needed
-	LastCheque *Cheque  // TODO: maybe at most just a serial number rather than full cheque
+	Beneficiary common.Address
```
Comment (mortelli) [inline swap/types.go:43]: We went over this with @zelig and it appears we only need the beneficiary address for now (which will eventually be assigned to the check object).

- `Peer` is not needed as it is implicit when receiving a message.
- `LastCheque` is not needed as the receiving node will check its own last known cheque.
- `PubKey` is not needed as it was only used to calculate the beneficiary address, which can be directly given as it is now.

File: swap/swap.go [hunk 2]
```diff
@@ -119,15 +123,36 @@ func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
...
+
 	//adjust the balance
 	//if amount is negative, it will decrease, otherwise increase
 	s.balances[peer.ID()] += amount
+
 	//save the new balance to the state store
 	peerBalance := s.balances[peer.ID()]
 	err = s.stateStore.Put(peer.ID().String(), &peerBalance)
+	if err != nil {
+		return
```
Comment (mortelli) [inline swap/swap.go]: It's better style to do explicit returns.

File: swap/swap.go [hunk 3]
```diff
@@ -119,15 +123,36 @@ func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
...
+	}
 
 	log.Debug(fmt.Sprintf("balance for peer %s: %s", peer.ID().String(), strconv.FormatInt(peerBalance, 10)))
-	return err
+
+	//check if balance with peer is over the payment threshold
```
Comment (mortelli) [inline swap/swap.go:146]: Look into a less spammy implementation.

File: swap/swap.go [hunk 4]
```diff
@@ -184,9 +209,10 @@ func (swap *Swap) Close() {
 // resetBalance is called:
 // * for the creditor: on cheque receival
 // * for the debitor: on confirmation receival
-func (s *Swap) resetBalance(peer enode.ID) {
+func (s *Swap) resetBalance(peerID enode.ID) {
 	//TODO: reset balance based on actual amount
-	s.balances[peer] = 0
+	//Lock() is not applied because it is expected to be done in the handleChequeRequestMsg call
+	s.balances[peerID] = 0
```
Comment (mortelli) [inline swap/swap.go:215]: Look into reducing the balance based on the cheque information instead of setting it back to 0.

Comment (vojtechsimetka) [reply swap/swap.go:215]: During our discussion we concluded it should be a cumulative value and compared as such. So in practice current balance would be retrieved as such:

`balance = balanceCummulative - lastChequeAmount/pricePerByte`

and the trigger for cheque request

`balance >= treshHold`

So this function should not exist.

File: swap/swap.go (hunk 3 above)
Comment (Eknir) [reply swap/swap.go:146]: I think that for now, we can leave it as is, but we should create an issue

Comment (Eknir) [reply swap/swap.go:146]: #1549

File: swap/swap.go (hunk 4 above)
Comment (Eknir) [reply swap/swap.go:215]: https://github.com/ethersphere/swarm/issues/1551

**Review by vojtechsimetka — APPROVED** [review]
I don't like the balance being reseted but maybe it is fine for the first approximation and new issue has been created #1551.

**Review by holisticode — CHANGES_REQUESTED** [review]
Looks good to me, thank you.

However, `TestRepeatedBookings` tests several things.
Could you think of some simple unit test which **only** would test thresholds?
I would like to see such a unit test.

**Pull request merged**